### PR TITLE
Rewrote browse APIs, support for backupset level browse

### PR DESCRIPTION
**Pull request by vaakash**
The browse function is now moved from subclient to backupset level.
It accepts dynamic arguments ( a dictionary of options or keyword arguments ). Default options for browse can be see in the backupset class init.

Redundant APIs like `browse_in_time` are removed as `browse` itself can be used.
With this change, browse now supports all the options which are supported in GUI including filters and can be easily extended for all requirements.

Return type of the browse function is unchanged.

### CODE REVIEW

File: cvpysdk/backupset.py [hunk 1]
```diff
@@ -663,6 +696,278 @@ def _update(self, backupset_name, backupset_description, default_backupset):
 
         return self._process_update_reponse(request_json)
 
+    @staticmethod
+    def _get_timestamp(timestamp):
+        """ Returns the timestamp which are in the format %Y-%m-%d %H:%M:%S
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: Please correct the Doc-string here.
The function is expecting the time in the format %Y-%m-%d %H:%M:%S, here,
and returns the Epoch time

Comment (vaakash) [inline cvpysdk/backupset.py]: Got it ! Missed to change the doc here..

**Comment by vaakash** [comment]
I have made a commit to support browse/find from instance level. Please review

**Comment by vaakash** [comment]
I have made a commit to support browse/find from instance level. Please review

**Comment by vaakash** [comment]
Any update on this PR ?

File: README.rst [hunk 2]
```diff
@@ -54,14 +54,23 @@ Run backup for a backupset:
...
 	>>> job = subclient.backup(backup_level, incremental_backup, incremental_level)
 
-Browsing content of a subclient:
-	>>> paths, dictionary = subclient.browse(path, show_deleted_files, vm_file_browse, vm_disk_browse)
+Browsing content at subclient level:
+	>>> paths, dictionary = subclient.browse(path = 'c:\\', show_deleted = True)
 
 Browsing content of a subclient in a specific time range:
-	>>> paths, dictionary = subclient.browse_in_time(path, show_deleted_files, restore_index, from_date, to_date)
+	>>> paths, dictionary = subclient.browse(path = 'f:\\', from_time = '2010-04-19 02:30:00', to_time = '2014-12-20 12:00:00')
```
Comment (GoelJatin) [inline README.rst]: Please remove the spaces before and after '=' as per PEP8

File: cvpysdk/backupset.py [hunk 3]
```diff
@@ -494,6 +511,22 @@ def __init__(self, instance_object, backupset_name, backupset_id=None):
         self.subclients = Subclients(self)
         self.schedules = Schedules(self)
 
+        self.default_browse_options = {
+            'operation': 'browse',
+            'show_deleted': False,
+            'from_time': 0,  # Timestamp or %Y-%m-%d %H:%M:%S
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: Isn't timestamp actually the string that you're providing in or

File: cvpysdk/backupset.py [hunk 4]
```diff
@@ -663,6 +696,278 @@ def _update(self, backupset_name, backupset_description, default_backupset):
 
         return self._process_update_reponse(request_json)
 
+    @staticmethod
+    def _get_timestamp(formatted_time):
+        """ Returns the epoch timestamp given the input time is in format %Y-%m-%d %H:%M:%S
+        
+        Args:
+            formatted_time          (multiple)  --  String of formatted time or integer of timestamp
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: multiple is not python data type. Please use correct data type here.

File: cvpysdk/backupset.py [hunk 5]
```diff
@@ -663,6 +696,278 @@ def _update(self, backupset_name, backupset_description, default_backupset):
...
 
+    @staticmethod
+    def _get_timestamp(formatted_time):
+        """ Returns the epoch timestamp given the input time is in format %Y-%m-%d %H:%M:%S
+        
+        Args:
+            formatted_time          (multiple)  --  String of formatted time or integer of timestamp
+            
+        Returns:
+            (int)- Timestamp
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: No parenthesis should come around the return data type
Also, elaborate more on the return data

Follow same style as per the design

File: cvpysdk/backupset.py [hunk 6]
```diff
@@ -663,6 +696,278 @@ def _update(self, backupset_name, backupset_description, default_backupset):
...
+            try:
+                return int(time.mktime(datetime.datetime.strptime(formatted_time, "%Y-%m-%d %H:%M:%S").timetuple()))
+            except:
+                raise SDKException('Subclient', '106')
+
+    def _set_defaults(self, custom, defaults):
+        """ Recursively sets default values on a dictionary
+
+        Args:
+            custom                  (dict)    --  The dictionary which has to be set with defaults
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: So much spacing b/w the arg, and its type is not required

File: cvpysdk/backupset.py [hunk 7]
```diff
@@ -663,6 +696,278 @@ def _update(self, backupset_name, backupset_description, default_backupset):
...
+            'find': 1
+        }
+
+        options['operation'] = options['operation'].lower()
+        options['operation'] = 'find' if options['operation'] not in operation_types else options['operation']
+
+        request_json = {
+            "opType": operation_types[options['operation']],
+            "mode": {
+                "mode": 2
```
Comment (GoelJatin) [inline cvpysdk/backupset.py]: The mode cannot be set to 2 by default, as some agents are not able to browse with mode 2.
For some agents, the browse mode is 4

File: cvpysdk/instance.py [hunk 8]
```diff
@@ -302,3 +302,93 @@ def instance_id(self):
...
+            SDKException:
+                If there are more than one backupsets in the instance 
+        
+        """
+
+        all_backupsets = self.backupsets._get_backupsets()
+
+        # If there is only one backupset, then perform backupset level browse on it
+        if len(all_backupsets.keys()) == 1:
+            the_lone_backupset = all_backupsets.keys()[0]
```
Comment (GoelJatin) [inline cvpysdk/instance.py]: Please correct the naming here. Use proper variable names

File: cvpysdk/instance.py [hunk 9]
```diff
@@ -302,3 +302,93 @@ def instance_id(self):
...
+            (list): List of only the file, folder paths from the browse response
+            (dict): Dictionary of all the paths with additional metadata which are retrieved from browse
+
+        """
+
+        all_backupsets = self.backupsets._get_backupsets()
+
+        # If there is only one backupset, then perform backupset level find on it
+        if len(all_backupsets.keys()) == 1:
+            the_lone_backupset = all_backupsets.keys()[0]
```
Comment (GoelJatin) [inline cvpysdk/instance.py]: Use proper variable names.

File: cvpysdk/backupset.py (hunk 3 above)
Comment (vaakash) [inline cvpysdk/backupset.py]: Timestamp provided can be either string or int.
If it is formatted time, then it should be a string

File: README.rst (hunk 2 above)
Comment (vaakash) [inline README.rst]: Please refer https://www.python.org/dev/peps/pep-0008/#other-recommendations

File: cvpysdk/backupset.py (hunk 7 above)
Comment (vaakash) [inline cvpysdk/backupset.py]: Mode does not play a major role in browse request. I tested for VSA with mode 2 and it worked. I am not making any changes here.

**Comment by vaakash** [comment]
Made the changes as requested.

File: cvpysdk/backupset.py (hunk 7 above)
Comment (GoelJatin) [inline cvpysdk/backupset.py]: I just tried with mode 2, with the code that you shared, and it didn't work for me.

Can you please check again.

File: cvpysdk/subclients/vssubclient.py [hunk 10]
```diff
@@ -240,359 +228,135 @@ def _process_restore_request(self, vm_names, restore_content):
...
-                    if failed to browse content
-
-                    if response is empty
-
-                    if response is not success
-        """
-        return self.browse_in_time(
-            vm_path, show_deleted_files, restore_index, True, from_date, to_date
-        )
+        return self.browse(*args, **kwargs)
```
Comment (GoelJatin) [inline cvpysdk/subclients/vssubclient.py]: Before browse, the value for vm_disk_browse, should be set to True in the options Dict

File: cvpysdk/subclients/vssubclient.py [hunk 11]
```diff
@@ -240,359 +228,135 @@ def _process_restore_request(self, vm_names, restore_content):
...
-                    if failed to browse content
-
-                    if response is empty
-
-                    if response is not success
-        """
-        browse_content = self.browse_in_time(
-            vm_path, show_deleted_files, restore_index, True, from_date, to_date
-        )
+        browse_content = self.browse(*args, **kwargs)
```
Comment (GoelJatin) [inline cvpysdk/subclients/vssubclient.py]: Before browse, the value for vm_disk_browse, should be set to True in the options Dict

**Pull request merged**